### PR TITLE
Port the code for HandleEvent from the java sdk

### PR DESCRIPTION
**Pull request by marto83**

### CODE REVIEW

File: src/IO.Ably.Shared/Push/ActivationStateMachine.cs [hunk 1]
```diff
@@ -61,9 +68,73 @@ private async Task ValidateRegistration()
...
+            finally
+            {
+                _handlingEvent = false;
+            }
+        }
+
+        private void PersistState()
         {
-            throw new NotImplementedException();
+            if (CurrentState != null && CurrentState.Persist)
```
Comment (tomkirbygreen) [inline src/IO.Ably.Shared/Push/ActivationStateMachine.cs:127]: Consider: Is it worth having something like:

```
var sharedName = PersistKeys.StateMachine.SharedName;
```

At the top level of the function so we that can simplify the three use cases and avoid a bit if reference-chasing?

**Review by tomkirbygreen — COMMENTED** [review]
Just a minor point, otherwise good.

File: src/IO.Ably.Shared/Push/ActivationStateMachine.cs (hunk 1 above)
Comment (marto83) [reply src/IO.Ably.Shared/Push/ActivationStateMachine.cs:127]: I'll get it updated in the next PR if you don't mind :)

File: src/IO.Ably.Shared/Push/ActivationStateMachine.cs [hunk 2]
```diff
@@ -61,9 +68,73 @@ private async Task ValidateRegistration()
             }
         }
 
-        private async Task HandleEvent(Event @event)
+        public async Task HandleEvent(Event @event)
+        {
+            lock (_handleEventsLock)
```
Comment (sacOO7) [inline src/IO.Ably.Shared/Push/ActivationStateMachine.cs:73]: @marto83 I thought we were not using any locks in the code

Comment (marto83) [reply src/IO.Ably.Shared/Push/ActivationStateMachine.cs:73]: fair point. I'm currently just porting the code from Java to get it working and when I get both Android and iOS working, I'll refactor it to use our workflow process. 
I'm currently porting and learning how it works.

**Pull request merged**